### DISCUSSION TITLE
Multithreading - Graphics 5 or Graphics 4 ?

### DISCUSSION THREAD

**Discussion by gitindro** [Kode/Kore#784] [General] [answered]
First of all thank you for this awesome library. 

I saw that g5 has command lists so I am guessing if I want a multi-threaded renderer where I want to populate render command from other threads g5 will make it easier, am I correct in my assumption? . Also, are there any other differences between them that I should know? My target harware are mostly mobile devices.

Thanks,

**Comment by Apprentice-Alchemist** [comment]
> I saw that g5 has command lists so I am guessing if I want a multi-threaded renderer where I want to populate render command from other threads g5 will make it easier, am I correct in my assumption?

Yes but actually no.
G5 is indeed intended to be a "modern" api that supports multithreading, but as far as I know none of the backends are thread-safe at the moment.
So right now G5 doesn't really have any advantages compared with G4.

Specifically for Android you'll also want to use the OpenGL backend, because despite being the default the Vulkan backend still has some bugs.

**Reply by gitindro** [reply]
got it thanks.

**Comment by RobDangerous** [comment, marked as answer]
G5 has a very big advantage because buffer-handling is not automatic like it is in G4 - you have to manually make sure that you don't write to buffers while the GPU uses them but that also makes performance much more reliable. The fully automatic buffer-handling in D3D11 and OpenGL is their biggest performance-issue in my experience.
And multithreaded command-list-recording does indeed still have a problem in Vulkan and D3D12. It's fine in Metal though as far as I know.
And @Apprentice-Alchemist, if you know about Vulkan-bugs, please file issues for them. Currently I see only #783 (which is already half-fixed).

**Reply by Apprentice-Alchemist** [reply]
Good point about buffer-handling, hadn't thought of that.

> It's fine in Metal though as far as I know

https://github.com/Kode/Kinc/blob/d26f103a34c5f4e503a98f23e75c0e20fa3733bd/Backends/Graphics5/Metal/Sources/kinc/backend/graphics5/commandlist.m.h#L30-L31
Doesn't look thread-safe to me.

**Reply by RobDangerous** [reply]
Right, overlooked those. That's all in #785 now.
So what are those other Vulkan-bugs you hinted at?

**Reply by gitindro** [reply]
Thanks, understood, I think I can continue using g4 then, dont really have much experience with mordern apis, have not done much at low level since dx 9 days.

**Reply by RobDangerous** [reply]
The intended way of doing things is to just start with G4 and migrate to G5 once you need it. Multithreaded-command-lists become important once you are CPU-bound. Other performance-problems (like the aforementioned buffer-handling) typically hit you earlier though.

**Reply by RobDangerous** [reply]
*once you are CPU-bound from command-list recording I mean. Plenty of other opportunities for becoming CPU-bound of course.
